middleware/rate-limiter.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+const {
+    RateLimiterMySQL,
+    RateLimiterMemory
+} = require('rate-limiter-flexible');
+const { startsWith } = require('lodash');
+
+const { sequelize } = require('../db/models/index');
+const appData = require('../common/appData');
+const logger = require('../common/logger').child({
+    service: 'rate-limiter'
+});
+
+const sqlDialect = startsWith(process.env.DB_CONNECTION_URI, 'sqlite://')
+    ? 'sqlite'
+    : 'mysql';
+
+const rateLimiterConfigs = {
+    failByUsername: {
+        maxPoints: 5,
+        keyPrefix: 'login_fail_consecutive_username',
+        duration: 60 * 60 * 3, // Points will expire after 3 hours
+        blockDuration: 60 * 15 // Exceeding points locks user out for 15 minutes
+    }
+};
+
+// We have to create all the ratelimiter db instances here
+// so the tables exist and are ready when the middleware is used.
+// We also need to be able to access the maxPoints
+// within the class below (which this API doesn't expose)
+// so we set it here as a property against the ratelimiter object
+let availableRateLimiters = {};
```

Comment:
Couple of options to simplify this, you could either use lodash `mapValues` to map over this like

```js
const availableRateLimiters = mapValues(rateLimiterConfigs, (conf, name) => {
    return {
        id: name,
        config: conf,
        instance:
            sqlDialect === 'mysql'
                ? new RateLimiterMySQL({
                      storeClient: sequelize,
                      dbName: `rate_limiter_${appData.environment}`,
                      tableName: 'rate_limiter',
                      keyPrefix: conf.keyPrefix,
                      points: conf.maxPoints,
                      duration: conf.duration,
                      blockDuration: conf.blockDuration
                  })
                : new RateLimiterMemory({
                      keyPrefix: conf.keyPrefix,
                      points: conf.maxPoints,
                      duration: conf.duration,
                      blockDuration: conf.blockDuration
                  })
    };
});
```

Or, perhaps make this a function instead:

```js
function rateLimiterFor(conf) {
    if (sqlDialect === 'mysql') {
        return new RateLimiterMySQL({
            storeClient: sequelize,
            dbName: `rate_limiter_${appData.environment}`,
            tableName: 'rate_limiter',
            keyPrefix: conf.keyPrefix,
            points: conf.maxPoints,
            duration: conf.duration,
            blockDuration: conf.blockDuration
        });
    } else {
        new RateLimiterMemory({
            keyPrefix: conf.keyPrefix,
            points: conf.maxPoints,
            duration: conf.duration,
            blockDuration: conf.blockDuration
        });
    }
}

function makeRateLimiter(id, conf) {
    return {
        id: id,
        config: conf,
        instance: rateLimiterFor(conf)
    };
}

const rateLimiterConfigs = {
    failByUsername: makeRateLimiter('failByUsername', {
        maxPoints: 5,
        keyPrefix: 'login_fail_consecutive_username',
        duration: 60 * 60 * 3, // Points will expire after 3 hours
        blockDuration: 60 * 15 // Exceeding points locks user out for 15 minutes
    })
};
```

or some combination of the two.